### ISSUE TITLE
How to Use Multi-ID 

### ISSUE THREAD

**Issue by GavinZhao19** [InstantID/InstantID#81]
Hello,

I recently came across the Multi-ID feature mentioned in the documentation, but I'm having trouble finding clear examples of how to use it. Specifically, I'm interested in combining images of different individuals, such as uploading two separate photos and generating a composite image.

For instance, I'd like to upload photos of two celebrities, Yang Mi and Taylor Swift, and create a merged image of them. However, I haven't found any examples in the code or on the Hugging Face platform that demonstrate how to achieve this.

Could you please provide some guidance or instructions on how to use the Multi-ID feature for this purpose? Any example code or detailed steps would be greatly appreciated.

Thank you for your assistance.

**Comment by ResearcherXman** [comment]
We don't have a specific plan to release this part of codes by now, but we do encourage the community to implement and PR.

**Comment by zewolf5** [comment]
Well. Since the answer to having multiple input images is suggested here and works:
https://github.com/InstantID/InstantID/issues/56#issuecomment-1908653153
> > Not sure if its already there, but multiple reference images? I could not find a way to select more than one image in the sample code provided. Topic #33 mentioned use of multiple reference images for better results.
> 
> I'm no expert but this should work `image = pipe( ... image_embeds=(face_emb1 + face_emb2 + face_emb3 + face_emb4)/4, ...) `

It seems to also work for adding different pictures into the same calculation. Did some quick testing and it seems to be the way to go.

Load the different pictures, either many from only one or multiple people and get a combined "face".
If you have gotten the sample code up and running, you got something like

`face_emb = face_info['embedding']
`
change it to
`face_emb = (face_info1['embedding'] + face_info2['embedding']) / 2`
or
`face_emb = (face_info1['embedding'] + face_info2['embedding'] + face_info3['embedding']) / 3`
etc.

I made a function for that taking a folder of face images:
`face_embeds = helpers.getAverageFaceEmb(face_folder)`

So get 1 face_info object for each image. Add and divide. I myself have not quite read and understood the concepts, but seems to be some math vectors or such that defines what a face looks like. So pure math to add the definition of the faces.

**Comment by platform-kit** [comment]
> We don't have a specific plan to release this part of codes by now, but we do encourage the community to implement and PR.

Is there a reason you are being vague about it? What do you gain by not releasing the code that you demonstrated in your paper?

**Comment by wangqixun** [comment]
> > We don't have a specific plan to release this part of codes by now, but we do encourage the community to implement and PR.
> 
> Is there a reason you are being vague about it? What do you gain by not releasing the code that you demonstrated in your paper?

maybe its name is not OpenAID

**Comment by usamaehsan** [comment]
> Well. Since the answer to having multiple input images is suggested here and works: [#56 (comment)](https://github.com/InstantID/InstantID/issues/56#issuecomment-1908653153)
> 
> > > Not sure if its already there, but multiple reference images? I could not find a way to select more than one image in the sample code provided. Topic #33 mentioned use of multiple reference images for better results.
> > 
> > 
> > I'm no expert but this should work `image = pipe( ... image_embeds=(face_emb1 + face_emb2 + face_emb3 + face_emb4)/4, ...) `
> 
> It seems to also work for adding different pictures into the same calculation. Did some quick testing and it seems to be the way to go.
> 
> Load the different pictures, either many from only one or multiple people and get a combined "face". If you have gotten the sample code up and running, you got something like
> 
> `face_emb = face_info['embedding'] ` change it to `face_emb = (face_info1['embedding'] + face_info2['embedding']) / 2` or `face_emb = (face_info1['embedding'] + face_info2['embedding'] + face_info3['embedding']) / 3` etc.
> 
> I made a function for that taking a folder of face images: `face_embeds = helpers.getAverageFaceEmb(face_folder)`
> 
> So get 1 face_info object for each image. Add and divide. I myself have not quite read and understood the concepts, but seems to be some math vectors or such that defines what a face looks like. So pure math to add the definition of the faces.

@zewolf5 can we use different weight to each image?

**Comment by G-force78** [comment]
> We don't have a specific plan to release this part of codes by now, but we do encourage the community to implement and PR.

In your original demo on hf you had it implemented. Why was it removed?

**Comment by cubiq** [comment]
It's possible with attention masking

![ComfyUI_temp_shjmq_00009_](https://github.com/InstantID/InstantID/assets/427614/01dff5da-9471-4e2c-9dd0-b475571d8f69)

(this is done in comfyui)

![image](https://github.com/InstantID/InstantID/assets/427614/9c3f8b54-5282-4310-a8fc-11f868450912)

https://github.com/cubiq/ComfyUI_InstantID

**Comment by Enzzer** [comment]
> It's possible with attention masking
> 
> ![ComfyUI_temp_shjmq_00009_](https://private-user-images.githubusercontent.com/427614/303904697-01dff5da-9471-4e2c-9dd0-b475571d8f69.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc4MTkyMjYsIm5iZiI6MTcwNzgxODkyNiwicGF0aCI6Ii80Mjc2MTQvMzAzOTA0Njk3LTAxZGZmNWRhLTk0NzEtNGUyYy05ZGQwLWI0NzU1NzFkOGY2OS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxM1QxMDA4NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZDg2OGE3MGIwNzk0ZTY0YzFkYWY0YTE5MDljZDNhZTA0Zjg5MWRhNTAwYjgwODg5YTU0ZWU5Y2U2MzlhNTZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.5ZwEI0ciRg9c3aSCmH3VqmUmV1oYNwT9UX52WHQ5wTI)
> 
> (this is done in comfyui)
> 
> ![image](https://private-user-images.githubusercontent.com/427614/303904851-9c3f8b54-5282-4310-a8fc-11f868450912.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc4MTkyMjYsIm5iZiI6MTcwNzgxODkyNiwicGF0aCI6Ii80Mjc2MTQvMzAzOTA0ODUxLTljM2Y4YjU0LTUyODItNDMxMC1hOGZjLTExZjg2ODQ1MDkxMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIxM1QxMDA4NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMTllMDc0YzYxMTFjNTlkZDlkY2QwZWQ1YWY1ZWI4ZWYwOTczNWMxZGFlM2ZjYThkNDFmZTQ5ZmUyYWMyZjBiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.va8UxT5q-chnrS3n4Ksm9YiL0JFcMy9VNRJGsxylfVM)
> 
> https://github.com/cubiq/ComfyUI_InstantID

Can you please share the workflow you used to do this?

**Comment by cubiq** [comment]
@Enzzer the workflow is in the repository

**Comment by Enzzer** [comment]
@cubiq Thanks for sharing the workflow!
I tried using it as it is (other than using my face and pose images) but I'm not getting the same results as you.
I'm getting both faces as the last input face image, seems like the first face is lost in the conditioning chain.
Any idea why?

**Comment by cubiq** [comment]
you probably just need to update the extension

**Comment by Enzzer** [comment]
@cubiq I tried updating both ComfyUI and your extension but I'm still getting the same face twice.
Can you share your input images so I can test with them?

**Comment by toonverbeek** [comment]
FWIW, also getting the same result as @Enzzer where both images are merged into one face. @Enzzer did you find a solution by any chance?

**Comment by cubiq** [comment]
the latest version, with the provided workflow works very well.

**Comment by toonverbeek** [comment]
Yes confirmed, I used the reference pose image from the repository with two custom portrait photo's and results are now a lot better.  Thanks for the update 👍

**Comment by xpluscal** [comment]
hey @toonverbeek can you share a workflow on how you combined pose with multi ID? How did you achieve this?

**Comment by eniora** [comment]
I know it's too late for this but I added it in my fork just today after many failed attempts (and help from AI chatbots):
https://github.com/eniora/InstantID-Unlocked
It works really well
